### DISCUSSION TITLE
How to update a subresource like "status" with KubernetesClient?

### DISCUSSION THREAD

**Discussion by jefflill** [kubernetes-client/csharp#882] [Q&A] [answered]
I'd like to update a custom resource with `spec` and `status` properties with the KubernetesClient.  What's the official way to do this?

* I've tried using `PatchClusterCustomObjectAsync()` but that didn't update the `status` on the API server.  I expected that because Kubernetes is supposed to treat the `spec` and `status` and properties separately to avoid triggering a watch when just the status changes.

* I don't see an API for updating just a subresource.  I tried appending `spec` to the object name like `myobject/spec` and then calling `PatchClusterCustomObjectAsync()`but that didn't work either.  I was hoping that `myobject/spec` would be inserted in the URI to reference the spec.

* I couldn't find any documentation or examples discussing this scenario, other than this two year old issue: https://github.com/kubernetes-client/csharp/issues/391

Follow up 1:
---

As seems typical, shortly after I submitted this, I realized I had a bug and I was never actually calling `PatchClusterCustomObjectAsync()`, so I'm hoping passing the object name like `myobject/spec` is going to work.

Follow up 2:
---

So I fixed my other issues and the name hack (`myobject/spec`) isn't working.  The problem is that the **"status"** part doesn't make it into the request URI.  I suspect that I'll have to craft my own HTTP request rather than using the Kubernetes API.

Question: **Can the KubernetesClient handle subresource updating?**

**Comment by tg123** [comment]
could you please take a look at https://github.com/kubernetes-client/csharp/tree/master/examples/customResource
i did not understand why spec cannot be patched, any err?

**Reply by jefflill** [reply]
I can patch the **spec** just fine.  My problem is updating the **status**.

There is no error but the **status** doesn't update.  It looks like **status** updates are special.  Here's how this works for the GO client:

https://book-v1.book.kubebuilder.io/basics/status_subresource.html

I wasn't able to figure out how to do this for for the .NET client.  It appears that **status** is a subresource that is directly under the main custom resource.  So if the URI for a cluster scoped custom resource is:

```
/apis/GROUP/VERSION/RESOURCETYPE/myresource
```

then it's status will be referenced via:

```
/apis/GROUP/VERSION/RESOURCETYPE/myresource/status
```

It also looks like its possible to specify different RBAC role rules that can manage permissions to the object separately from the status, so some apps could be restricted from managing the object and other apps could be restricted from managing the status.

So my question is: **Can I use the client API to update the status?**

**Reply by tg123** [reply]
see `client.CustomObjects.PatchClusterCustomObjectStatus`
is it what you are looking for?

I hope the api group helps you find the api well and i am going to create a doc site when I have time for easier api lookup

**Reply by tg123** [reply]
`CustomObjects.PatchClusterCustomObjectStatus` as well

**Comment by jefflill** [comment, marked as answer]
@tg123: Thanks!!!  This is what I was looking for:
```
client.CustomObjects.PatchClusterCustomObjectStatus()
```
I don't know how I didn't find this myself.

You might want to cover these topics when you do your write up:

* RBAC role: be sure to add a rule that allows **patch** for **MYRESOURCE/status**
* I first thought that the patch was rooted under the status property, but its actually rooted top of the document object.

**Reply by tg123** [reply]
happy to know it, could you please sent a PR to add `update status` to the customResource example?
